testdata/p4_16_errors/typedef-and-type-definitions.p4

```diff
@@ -0,0 +1,158 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <core.p4>
+
+
+header h1_t {
+    bit<8> f1;
+}
+
+header h2_t {
+    bit<8> f2;
+}
+
+header_union hu1_t {
+    h1_t h1;
+    h2_t h2;
+}
+
+struct s1_t {
+    bit<8> f2;
+}
+
+enum enum1_t {
+    A,
+    B
+}
+
+enum bit<7> serializable_enum1_t {
+    C = 3,
+    D = 8
+}
+
+// Naming convention used here:
+
+// D is abbreviation for `typedef`
+// T is abbrevation for `type`
+
+// Since types and typedefs can be defined in terms of each other, the
+// names I use here contain sequences of Ds and Ts to indicate the
+// order in which they have been "stacked", e.g. EthDT_t is a `type`
+// (the T is last) defined on type of a `typedef` (the D just before
+// the T).
+
+
+//////////////////////////////////////////////////////////////////////
+// typdef
+//////////////////////////////////////////////////////////////////////
+
+// int, void, match_kind cause syntax error that causes compiler to
+// stop immediately without looking for further errors.
+
+// syntax error, unexpected IDENTIFIER, expecting <
+//typedef int        intD_t;
+// syntax error, unexpected VOID, expecting ENUM or HEADER or HEADER_UNION or STRUCT
+//typedef void       voidD_t;
+//syntax error, unexpected MATCH_KIND, expecting ENUM or HEADER or HEADER_UNION or STRUCT
+//typedef match_kind matchD_t;
+
+// The compiler allows all of the typedef definitions below, without
+// error.
+
+typedef bit<8>     bD_t;
+typedef int<8>     iD_t;
+typedef varbit<8>  vD_t;
+typedef error      errorD_t;
+typedef bool       boolD_t;
+typedef enum1_t    enum1D_t;
+typedef serializable_enum1_t serializable_enum1D_t;
+typedef h1_t       H1D_t;
+typedef header h2a_t { bit<16> f2; bit<8> f3; } H2aD_t;
+typedef h1_t[3]    h1StackD_t;
+typedef hu1_t      hu1D_t;
+typedef s1_t       s1D_t;
+typedef struct PointA_t { int<32> x; int<32> y; } PointAD_t;
+typedef tuple<bit<8>, bit<17> > Tuple1D_t;
+typedef tuple<bit<8>, varbit<17> > Tuple2D_t;
+
+
+//////////////////////////////////////////////////////////////////////
+// type
+//////////////////////////////////////////////////////////////////////
+
+// int, void, match_kind cause syntax error that causes compiler to
+// stop immediately without looking for further errors.
+
+// syntax error, unexpected IDENTIFIER, expecting <
+//type    int        intT_t;
+// syntax error, unexpected VOID, expecting ENUM or HEADER or HEADER_UNION or STRUCT
+//type    void       voidT_t;
+// syntax error, unexpected MATCH_KIND, expecting ENUM or HEADER or HEADER_UNION or STRUCT
+//type    match_kind matchT_t;
+
+
+// Every line marked '// error' below causes the compiler to issue an
+// error message of this form:
+
+// error: <type_name>: 'type' can only be applied to base types or tuple types
```

Comment:
The spec does not seem to say much about restrictions on type. that looks like a bug. Do we really want to allow tuple?

Reply:
From the error message it looks like we do want to allow tuples, but I don't remember what the use cases for tuples are.

Reply:
I do not recall reading any description of why tuples are (or should be) supported for `type` definitions, but did notice them mentioned in the error message explicitly like that.

I am not aware of any reasons for the PSA architecture's use of `type` definitions, for runtime translation, that it needs support for `type` on top of a tuple type.  So far it only needs `type` on top of a `bit<W>` type.